pkg/vulnsrc/openeuler/openeuler.go

```diff
@@ -0,0 +1,267 @@
+}
+
+func (vs VulnSrc) Update(dir string) error {
+	log.Println("Saving openEuler CVRF")
+	var cvrfs []Cvrf
+	rootDir := filepath.Join(dir, "vuln-list", eulerDir)
+	err := utils.FileWalk(rootDir, func(r io.Reader, path string) error {
+		var cvrf Cvrf
+		if err := json.NewDecoder(r).Decode(&cvrf); err != nil {
+			return xerrors.Errorf("failed to decode openEuler CVRF JSON: %w %+v", err, cvrf)
+		}
+		cvrfs = append(cvrfs, cvrf)
+		return nil
+	})
+	if err != nil {
+		return xerrors.Errorf("error in openEuler CVRF walk: %w", err)
+	}
+
+	if err = vs.save(cvrfs); err != nil {
+		return xerrors.Errorf("error in openEuler CVRF save: %w", err)
+	}
+
+	return nil
+}
+
+func (vs VulnSrc) save(cvrfs []Cvrf) error {
+	err := vs.dbc.BatchUpdate(func(tx *bolt.Tx) error {
+		return vs.commit(tx, cvrfs)
+	})
+	if err != nil {
+		return xerrors.Errorf("error in batch update: %w", err)
+	}
+	return nil
+}
+
+func (vs VulnSrc) commit(tx *bolt.Tx, cvrfs []Cvrf) error {
+	var uniqOSVers = make(map[string]struct{})
+	for _, cvrf := range cvrfs {
+		affectedPkgs := getAffectedPackages(cvrf.ProductTree)
+		if len(affectedPkgs) == 0 {
+			continue
+		}
+
+		for _, pkg := range affectedPkgs {
+			arches := lo.Uniq(pkg.Archs)
+			sort.Strings(arches)
+			advisory := types.Advisory{
+				FixedVersion: pkg.FixedVersion,
+				Arches:       arches,
+			}
+			// Don't put the same data source multiple times.
+			if _, ok := uniqOSVers[pkg.OSVer]; !ok {
+				uniqOSVers[pkg.OSVer] = struct{}{}
+				if err := vs.dbc.PutDataSource(tx, pkg.OSVer, source); err != nil {
+					return xerrors.Errorf("failed to put data source: %w", err)
+				}
+			}
+
+			if err := vs.dbc.PutAdvisoryDetail(tx, cvrf.Tracking.ID, pkg.Name,
+				[]string{pkg.OSVer}, advisory); err != nil {
+				return xerrors.Errorf("unable to save %s CVRF: %w", pkg.OSVer, err)
+			}
+		}
+
+		var references []string
+		for _, ref := range cvrf.References {
+			references = append(references, ref.URL)
+		}
+
+		severity := types.SeverityUnknown
+		for _, cvuln := range cvrf.Vulnerabilities {
+			for _, threat := range cvuln.Threats {
+				sev := severityFromThreat(threat.Severity)
+				if severity < sev {
+					severity = sev
+				}
+			}
+		}
+
+		vuln := types.VulnerabilityDetail{
+			References:  references,
+			Title:       cvrf.Title,
+			Description: getDetail(cvrf.Notes),
+			Severity:    severity,
+		}
+		if err := vs.dbc.PutVulnerabilityDetail(tx, cvrf.Tracking.ID, source.ID, vuln); err != nil {
+			return xerrors.Errorf("failed to save openEuler CVRF vulnerability: %w", err)
+		}
+
+		// for optimization
+		if err := vs.dbc.PutVulnerabilityID(tx, cvrf.Tracking.ID); err != nil {
+			return xerrors.Errorf("failed to save the vulnerability ID: %w", err)
+		}
+	}
+	return nil
+}
+
+func getAffectedPackages(productTree ProductTree) []Package {
+	var pkgs = make(map[string]Package) // pkgID => Package
+	for _, branch := range productTree.Branches {
+		// `src` pkgs are not installed in openEuler.
```

Comment:
We can get a source package name from RPMBDB.

```
$ docker run --rm -it openeuler/openeuler:20.03-lts 
[root@5462ae7535d3 /]# rpm -qa --qf "%{NAME} %{SOURCERPM}\n" | grep glibc
glibc-common glibc-2.28-36.oe1.src.rpm
glibc glibc-2.28-36.oe1.src.rpm
```

So, I'm wondering if a source package name and a list of binary package names are equivalent in terms of vulnerability detection. In other words, wouldn't it be enough to save only the source package name and the fixed version in the DB? It has a great benefit from the perspective of DB size.

Let's take an example. Currently, we save all binary names here, like `glibc-devel`, `nscd`, etc.
https://github.com/aquasecurity/vuln-list/blob/ecca778da16828205caa70ba6941582f70041104/openeuler/2021/openEuler-SA-2021-1013.json#L79-L210

I'm wondering if storing only `glibc` is enough.
https://github.com/aquasecurity/vuln-list/blob/ecca778da16828205caa70ba6941582f70041104/openeuler/2021/openEuler-SA-2021-1013.json#L231-L242

We do that for other distributions. If only the binary packages really affected by the vulnerability are listed, they should be preserved, but the current advisory appears to list all binary packages.

Reply:
I tried to find out this, the result is as the example shows above that the really affected package is only `glibc`. I will change the code, for `aarch64` and `x86_64` packages, only the binary like `glibc` will be preserved, others like `glibc-common` or `glibc-debuginfo` ... will not be preserved. Is that OK?

Reply:
> only the binary like glibc will be preserved, others like glibc-common or glibc-debuginfo ... will not be preserved. Is that OK?

I meant we may want to store only source packages. The source [glibc](https://github.com/aquasecurity/vuln-list/blob/ecca778da16828205caa70ba6941582f70041104/openeuler/2021/openEuler-SA-2021-1013.json#L228-L243) and the binary [glibc](https://github.com/aquasecurity/vuln-list/blob/ecca778da16828205caa70ba6941582f70041104/openeuler/2021/openEuler-SA-2021-1013.json#L195-L199) are technically different. In this example, by chance, there is a binary package name identical to the source package name, but there should also be examples where it does not exist.

How about storing the source package name and the affected architectures, since I think there are cases where x86_64 is affected but aarch64 is not?

Reply:
Yes, such as `Affected package: glibc` with `Arches: [x86_64, aarch64, noaarch]`?

Reply:
Exactly. That's my idea. But I'm not confident as I'm not familiar with openEuler. If it turns out we need binary package names in the future, we will be able to fix it without breaking. If you also don't find any issues, we can go with source names.

Reply:
I have a question, how to use stored `src` package names to match the installed packages while there is no binary package name identical to the source package name?

Reply:
I will take some time to analyze this

Reply:
After research and confirmed, I prefer to preserve all binary packages as what we are doing now.

The reasons are:
 1. Assume the only affected package is `glibc`, but `glibc` is installed by installing those binary packages (not the `src` package). We have to determine whether `glibc` is fixed by checking whether all these binary packages were updated. In other words, those binary packages are the ones with real vulnerabilities (I have confirmed this from openEuler community) .

    For example, https://www.openeuler.org/en/security/security-bulletins/detail/?id=openEuler-SA-2024-1845 (the binary packages are also the updated packages)

2. I found SUSE also regards all the updated packages as `affected packages`, 
For example, although the main fixed package in https://lists.opensuse.org/archives/list/security-announce@lists.opensuse.org/message/E5TLNZZQUO45PVXGT6S2YSICBQSLYZRH/ is `mbedtls`, all packages listed are regarded as `affected packages` by https://github.com/aquasecurity/trivy-db/blob/36e595ad5ee3c360d107b62114e0fda4ca10cd99/pkg/vulnsrc/suse-cvrf/suse-cvrf.go#L167

    Above `relationships` are stored in https://github.com/aquasecurity/vuln-list/blob/29b6594e980c0c164a022f14f3026e752408ef48/cvrf/suse/opensuse/2015/openSUSE-SU-2015-2257-1.json#L35

Reply:
I have already written about how to get a source package.
https://github.com/aquasecurity/trivy-db/pull/397#discussion_r1678894862

>We have to determine whether glibc is fixed by checking whether all these binary packages were updated.

Are you saying only one of the binary packages can be updated? Is it possible that `glibc-common` is the latest and `nscd` is not? If that is the case, please share how to reproduce it with a container image. I tried [this image](https://hub.docker.com/r/openeuler/openeuler), but `dnf install` showed an error.

It's different from other Red Hat-based distributions. You can see below.

```
[root@3b1d8987d0c8 /]# rpm -qa --qf "%{NAME} %{SOURCERPM}\n" | grep glibc
glibc-minimal-langpack glibc-2.28-236.el8_9.12.src.rpm
glibc glibc-2.28-236.el8_9.12.src.rpm
glibc-common glibc-2.28-236.el8_9.12.src.rpm
[root@3b1d8987d0c8 /]# yum -y install glibc-common
AlmaLinux 8 - BaseOS                                                                                                                                                                  1.1 MB/s | 4.1 MB     00:03
AlmaLinux 8 - AppStream                                                                                                                                                               2.5 MB/s | 9.8 MB     00:03
AlmaLinux 8 - Extras                                                                                                                                                                  4.7 kB/s |  12 kB     00:02
Package glibc-common-2.28-236.el8_9.12.aarch64 is already installed.
Dependencies resolved.
======================================================================================================================================================================================================================
 Package                                                     Architecture                                 Version                                                   Repository                                   Size
======================================================================================================================================================================================================================
Upgrading:
 glibc                                                       aarch64                                      2.28-251.el8_10.2                                         baseos                                      1.8 M
 glibc-common                                                aarch64                                      2.28-251.el8_10.2                                         baseos                                      1.0 M
 glibc-minimal-langpack                                      aarch64                                      2.28-251.el8_10.2                                         baseos                                       68 k
Installing weak dependencies:
 glibc-gconv-extra                                           aarch64                                      2.28-251.el8_10.2                                         baseos                                      1.8 M
 glibc-langpack-en                                           aarch64                                      2.28-251.el8_10.2                                         baseos                                      830 k

Transaction Summary
======================================================================================================================================================================================================================
Install  2 Packages
Upgrade  3 Packages

Total download size: 5.4 M
Downloading Packages:
(1/5): glibc-2.28-251.el8_10.2.aarch64.rpm                                                                                                                                            996 kB/s | 1.8 MB     00:01
(2/5): glibc-common-2.28-251.el8_10.2.aarch64.rpm                                                                                                                                     1.0 MB/s | 1.0 MB     00:01
(3/5): glibc-langpack-en-2.28-251.el8_10.2.aarch64.rpm                                                                                                                                256 kB/s | 830 kB     00:03
(4/5): glibc-gconv-extra-2.28-251.el8_10.2.aarch64.rpm                                                                                                                                529 kB/s | 1.8 MB     00:03
(5/5): glibc-minimal-langpack-2.28-251.el8_10.2.aarch64.rpm                                                                                                                           103 kB/s |  68 kB     00:00
----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------
Total                                                                                                                                                                                 1.3 MB/s | 5.4 MB     00:04
Running transaction check
Transaction check succeeded.
Running transaction test
Transaction test succeeded.
Running transaction
  Preparing        :                                                                                                                                                                                              1/1
  Installing       : glibc-langpack-en-2.28-251.el8_10.2.aarch64                                                                                                                                                  1/8
  Upgrading        : glibc-common-2.28-251.el8_10.2.aarch64                                                                                                                                                       2/8
  Running scriptlet: glibc-2.28-251.el8_10.2.aarch64                                                                                                                                                              3/8
  Upgrading        : glibc-2.28-251.el8_10.2.aarch64                                                                                                                                                              3/8
  Running scriptlet: glibc-2.28-251.el8_10.2.aarch64                                                                                                                                                              3/8
  Installing       : glibc-gconv-extra-2.28-251.el8_10.2.aarch64                                                                                                                                                  4/8
  Running scriptlet: glibc-gconv-extra-2.28-251.el8_10.2.aarch64                                                                                                                                                  4/8
  Upgrading        : glibc-minimal-langpack-2.28-251.el8_10.2.aarch64                                                                                                                                             5/8
  Cleanup          : glibc-2.28-236.el8_9.12.aarch64                                                                                                                                                              6/8
  Cleanup          : glibc-minimal-langpack-2.28-236.el8_9.12.aarch64                                                                                                                                             7/8
  Cleanup          : glibc-common-2.28-236.el8_9.12.aarch64                                                                                                                                                       8/8
  Running scriptlet: glibc-common-2.28-236.el8_9.12.aarch64                                                                                                                                                       8/8
  Running scriptlet: glibc-common-2.28-251.el8_10.2.aarch64                                                                                                                                                       8/8
  Verifying        : glibc-gconv-extra-2.28-251.el8_10.2.aarch64                                                                                                                                                  1/8
  Verifying        : glibc-langpack-en-2.28-251.el8_10.2.aarch64                                                                                                                                                  2/8
  Verifying        : glibc-2.28-251.el8_10.2.aarch64                                                                                                                                                              3/8
  Verifying        : glibc-2.28-236.el8_9.12.aarch64                                                                                                                                                              4/8
  Verifying        : glibc-common-2.28-251.el8_10.2.aarch64                                                                                                                                                       5/8
  Verifying        : glibc-common-2.28-236.el8_9.12.aarch64                                                                                                                                                       6/8
  Verifying        : glibc-minimal-langpack-2.28-251.el8_10.2.aarch64                                                                                                                                             7/8
  Verifying        : glibc-minimal-langpack-2.28-236.el8_9.12.aarch64                                                                                                                                             8/8

Upgraded:
  glibc-2.28-251.el8_10.2.aarch64                                glibc-common-2.28-251.el8_10.2.aarch64                                glibc-minimal-langpack-2.28-251.el8_10.2.aarch64
Installed:
  glibc-gconv-extra-2.28-251.el8_10.2.aarch64                                                               glibc-langpack-en-2.28-251.el8_10.2.aarch64

Complete!
[root@3b1d8987d0c8 /]# rpm -qa --qf "%{NAME} %{SOURCERPM}\n" | grep glibc
glibc-common glibc-2.28-251.el8_10.2.src.rpm
glibc-gconv-extra glibc-2.28-251.el8_10.2.src.rpm
glibc-langpack-en glibc-2.28-251.el8_10.2.src.rpm
glibc glibc-2.28-251.el8_10.2.src.rpm
glibc-minimal-langpack glibc-2.28-251.el8_10.2.src.rpm
```

When you update a binary package, `glibc-common`, all other binary packages under the source package `glibc` are also updated.

Even if it is possible, it doesn't matter because you can check the source package version of each binary package. in the above example, when you find `glibc-common`, you extract the source package `glibc-2.28-251` and use it for vulnerability detection. Next, you find `glibc-gconv-extra`  and extract `glibc-2.28-36` for scanning. Even if the version (2.28-36) is different from other packages (2.28-251), you can still correctly detect vulnerabilities with source packages.

Reply:
>2. I found SUSE also regards all the updated packages as affected packages,

If I remember correctly, SUSE didn't have source package names in CSAF. But anyway, SUSE is not well maintained compared to other distributions. You should not reference it.

Reply:
I think I can understand you now, let me refactor the code. Thank you very much!

Reply:
It's a bit complicated. Please feel free to ask any questions.

Reply:
I did some changes for string only source packages, please take a look. Thank you very much!

Reply:
@knqyf263 Could you please re-check the latest changes currently? And let me know how to push this forward.
Thank you very much!

Reply:
I'll take a look this week.

Reply:
Thank you!

Reply:
Since I'm on sick leave now, I probably won't have time this week.

@DmitriyLewen Can you please take a look if you have time? We discussed using source package names rather than binary package names for efficiency (see above conversation).

Reply:
@knqyf263 Thank you!
@DmitriyLewen Could please review those changes? Thanks!

Reply:
Hi, @DmitriyLewen, could you please check the changes currently? Thank you very much!

Reply:
Hi @DmitriyLewen @knqyf263, I'm still looking forward to your suggestions, please review the new changes as soon as you have time. 
Thank you very much!

Reply:
@wjunLu Please be patient. Maintainers have numerous other tasks and might be taking the summer vacation. It is good to remind maintainers, but please do that only once a week. There is not much point in notifying daily.

Reply:
oh, sorry for that

Reply:
LGTM.
I updated the logic a bit to get the arches by OS version - eee8cdb

This is necessary to avoid adding unaffected arches.
See the test in 751c88d